aiida/tools/importexport/archive/migrations/v01_to_v02.py

```diff
@@ -27,41 +29,54 @@ def migrate_v1_to_v2(metadata, data, *args):
     old_start = 'aiida.djsite'
     new_start = 'aiida.backends.djsite'
 
+    metadata = cache.get(
+        'metadata.json', cache.get('metadata.json', json.loads((folder / 'metadata.json').read_text('utf8')))
+    )
+
     verify_metadata_version(metadata, old_version)
     update_metadata(metadata, new_version)
 
-    def get_new_string(old_string):
-        """Replace the old module prefix with the new."""
-        if old_string.startswith(old_start):
-            return f'{new_start}{old_string[len(old_start):]}'
-
-        return old_string
-
-    def replace_requires(data):
-        """Replace the requires keys with new module path."""
-        if isinstance(data, dict):
-            new_data = {}
-            for key, value in data.items():
-                if key == 'requires' and value.startswith(old_start):
-                    new_data[key] = get_new_string(value)
-                else:
-                    new_data[key] = replace_requires(value)
-            return new_data
-
-        return data
+    data = cache.get('data.json', json.loads((folder / 'data.json').read_text('utf8')))
 
     for field in ['export_data']:
         for key in list(data[field]):
             if key.startswith(old_start):
-                new_key = get_new_string(key)
+                new_key = get_new_string(key, old_start, new_start)
                 data[field][new_key] = data[field][key]
                 del data[field][key]
 
     for field in ['unique_identifiers', 'all_fields_info']:
         for key in list(metadata[field].keys()):
             if key.startswith(old_start):
-                new_key = get_new_string(key)
+                new_key = get_new_string(key, old_start, new_start)
                 metadata[field][new_key] = metadata[field][key]
                 del metadata[field][key]
 
-    metadata['all_fields_info'] = replace_requires(metadata['all_fields_info'])
+    metadata['all_fields_info'] = replace_requires(metadata['all_fields_info'], old_start, new_start)
+
+    (folder / 'metadata.json').write_text(json.dumps(metadata), encoding='utf8')
```

Comment:
Why does this need to be written every time?

It seems to me that the most efficient way to do this would be to:
 * use a context manager
 * let the migration function read data when needed, which then causes read of json file via context manager
 * only update metadata/data in context manager (never write to disk during migration)
 * on exit (after all migrations), let context manager check which of metadata/data were updated; write only those back

Reply:
well firstly you are assuming that metadata.json and data.json will always be present. Once the archive format is changed that will certainly not be the case (and is the main point of this PR 😉).
When there is a migration which converts this data.json to the new format, this method will no longer work.
Similarly, once we don't have to, we will not want to hold all the data in memory until all migrations are complete.
Essentially we are sacrificing some performance, for a more general approach, which will eventually allow for a large reduction in memory usage.

Reply:
I think passing the migration functions the migrator would be a big hassle, especially for testing and modularity.
But you could pass them something like:

```python
class CacheFolder:
  """Caches writes."""
  def __init__(self, path: Path):
    self._path = path
    self._cache = {}
  @property
  def path(self):
    return self._path
  def flush(self):
     for path, val in self._cache.items():
        (self._path / path).write_text(val, encoding="utf8")
  def write_text(path, content):
    self._cache[path] = content
  def read_text(path):
    if path in self._cache:
       return self._cache[path]
    return (self._path / path).read_text("utf8")
  def __enter__(self):
     return self
  def __exit__(self, *args):
     self.flush()
```

what do you think about that?

Reply:
implemented in cbcaec6

Reply:
> what do you think about that?

Looks great!